### PR TITLE
Story/2911 - Refactoring with package links

### PR DESCRIPTION
**Pull request by PeterC10**
Take into account package links when refactoring pickings

### CODE REVIEW

File: addons/udes_stock/models/stock_location.py [hunk 1]
```diff
@@ -312,3 +312,11 @@ def get_available_stock_locations(self):
          their children) stock available for fulfilling orders. Should be
         overridden where necessary"""
         return self.env.ref("stock.stock_location_stock")
+
+    def get_supplied_location_is_child(self, location):
```
Comment (armandcela) [inline addons/udes_stock/models/stock_location.py:316]: Would suggest to transfer this in base models, we may need to check for user templates, products etc. Maybe the name should be more generic example `record_is_child(self, child) `

Comment (PeterC10) [reply addons/udes_stock/models/stock_location.py:316]: Good idea, not sure on the name though (I didn't like `get_supplied_location_is_child` either).

The problem with `record_is_child` is that it implies that we are just checking if the record in `self` is a child of something else. But instead we are checking that `self` is a parent of the supplied record.

Maybe `record_is_parent`?

File: addons/udes_stock_refactoring/models/stock_picking.py [hunk 2]
```diff
@@ -43,6 +41,43 @@ def _remove_misleading_values(self, **kwargs):
                 if current_value and current_value != value:
                     self[field] = False
 
+    @api.model
+    def _setup_refactored_picking_package_levels(self, picking, move_lines):
```
Comment (armandcela) [inline addons/udes_stock_refactoring/models/stock_picking.py:45]: From what i see you are not using self, and you are on stock picking class. I would suggest to define method:
`def _setup_refactored_picking_package_levels(self, move_lines):`  -- where self is the picking, maybe can add ensure one on self and below when you call to call like: `picking._setup_refactored_picking_package_levels(move_lines)`

File: addons/udes_stock/models/stock_location.py (hunk 1 above)
Comment (armandcela) [reply addons/udes_stock/models/stock_location.py:316]: I think you are better with names, definitely :)

File: addons/udes_stock_refactoring/models/stock_picking.py (hunk 2 above)
Comment (PeterC10) [reply addons/udes_stock_refactoring/models/stock_picking.py:45]: Good spot, problem with being lazy and copying `_new_picking_for_group`.  Will amend.

File: addons/udes_stock/models/stock_location.py (hunk 1 above)
Comment (PeterC10) [reply addons/udes_stock/models/stock_location.py:316]: I wish that were true, I've added some very curious names in the code!

Comment (PeterC10) [reply addons/udes_stock/models/stock_location.py:316]: I think... `record_is_child_of`... that way it is clear from the name that we are checking it is a child of a particular location. :)

File: addons/udes_stock_refactoring/models/stock_picking.py [hunk 3]
```diff
@@ -43,6 +41,43 @@ def _remove_misleading_values(self, **kwargs):
...
+
+        By setting the package levels as needed we avoid extra package levels potentially
+        being created later on via `_check_entire_pack`. This then also avoids the potential for
+        erroneous moves being created when `_generate_moves` is called on a newly generated
+        package level.
+        """
+        PackageLevel = self.env["stock.package_level"]
+
+        move_lines_with_package_levels = move_lines.filtered("package_level_id")
+        package_levels = move_lines_with_package_levels.mapped("package_level_id")
```
Comment (ws-mps) [inline addons/udes_stock_refactoring/models/stock_picking.py:63]: I don't think we need `mapped` for relational fields anymore (same for line 66).

Comment (PeterC10) [reply addons/udes_stock_refactoring/models/stock_picking.py]: Nope, old habits... fixed.

File: addons/udes_stock_refactoring/models/stock_picking.py (hunk 2 above)
Comment (PeterC10) [reply addons/udes_stock_refactoring/models/stock_picking.py]: Fixed.

File: addons/udes_stock/models/stock_location.py (hunk 1 above)
Comment (PeterC10) [reply addons/udes_stock/models/stock_location.py]: Went for `record_is_child_of_self` and added unit tests, please review.

**Review by armandcela — APPROVED** [review]

File: addons/udes_stock/models/stock_picking.py [hunk 4]
```diff
@@ -871,8 +871,12 @@ def validate_picking(self, create_backorder=False, force_validate=False):
             # Check if a backorder is required and can be created:
             # a) Multi users is enabled
             # b) There are incomplete move lines and create backorder is allowed
+            # c) There are incomplete moves (ie partially available) and create backorder is allowed
             multi_users_enabled = self.picking_type_id.u_multi_users_enabled
-            if self.move_line_ids.get_lines_incomplete():
+            if (
+                self.move_line_ids.get_lines_incomplete()
+                or self.move_lines.get_moves_with_quantity_todo()
```
Comment (LoKKeR96) [inline addons/udes_stock/models/stock_picking.py:878]: Wouldn't we only need get_moves_with_quantity_todo()  ?

Comment (PeterC10) [reply addons/udes_stock/models/stock_picking.py]: Fair point, if we have a move line in a picking and not the move then something elsewhere has gone very wrong and we're all going to be very sad. Have made the change.

**Review by LoKKeR96 — APPROVED** [review]
As you mentioned in one of the commits, long term we might need to review the tests but so far so good!

**Pull request merged**